File: cert-processor/src/main/java/org/sunbird/incredible/processor/views/SvgGenerator.java
package org.sunbird.incredible.processor.views;

import org.apache.commons.io.IOUtils;
import org.apache.commons.text.StringSubstitutor;
import org.sunbird.incredible.processor.store.LocalStore;
import org.sunbird.incredible.pojos.CertificateExtension;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SvgGenerator {

    private Logger logger = LoggerFactory.getLogger(SvgGenerator.class);
    private String svgTemplate;
    private String directory;
    private static Map<String, String> encoderMap = new HashMap<>();
    private static Map<String, String> cachedSvgTemplates = new HashMap<>();


    static {
        encoderMap.put("<", "%3C");
        encoderMap.put(">", "%3E");
        encoderMap.put("#", "%23");
        encoderMap.put("%", "%25");
        encoderMap.put("\"", "\'");
    }

    public SvgGenerator(String svgTemplate, String directory) {
        this.svgTemplate = svgTemplate;
        this.directory = directory;
    }

    public String generate(CertificateExtension certificateExtension, String encodedQrCode) throws IOException {
        String svgFileName = getSvgFileName();
        String svgContent;
        File file = new File(directory + svgFileName);
        if (!file.exists()) {
            logger.info("{} file does not exits , downloading", svgFileName);
            download(svgFileName);
        }
        if (!cachedSvgTemplates.containsKey(this.svgTemplate)) {
            svgContent = readSvgContent(file.getAbsolutePath());
            String encodedSvg = "data:image/svg+xml," + encodeData(svgContent);
            encodedSvg = encodedSvg.replaceAll("\n", "").replaceAll("\t", "");
            cachedSvgTemplates.put(this.svgTemplate, encodedSvg);
        }
        logger.info("svg template is cached {}", cachedSvgTemplates.containsKey(this.svgTemplate));
        String svgData = replaceTemplateVars(cachedSvgTemplates.get(this.svgTemplate), certificateExtension, encodedQrCode);
        logger.info("svg template string creation completed");
        return svgData;
    }


    private String replaceTemplateVars(String svgContent, CertificateExtension certificateExtension, String encodeQrCode) {
        HTMLVarResolver htmlVarResolver = new HTMLVarResolver(certificateExtension);
        Map<String, String> certData = htmlVarResolver.getCertMetaData();
        certData.put("qrCodeImage", "data:image/png;base64," + encodeQrCode);
        StringSubstitutor sub = new StringSubstitutor(certData);
        String resolvedString = sub.replace(svgContent);
        logger.info("replacing temp vars completed");
        return resolvedString;
    }

    private String encodeData(String data) {
        StringBuffer stringBuffer = new StringBuffer();
        Pattern pattern = Pattern.compile("[<>#%\"]");
        Matcher matcher = pattern.matcher(data);
        while (matcher.find()) {
            matcher.appendReplacement(stringBuffer, encoderMap.get(matcher.group()));
        }
        matcher.appendTail(stringBuffer);
        return stringBuffer.toString();
    }

    private String readSvgContent(String path) throws IOException {
        FileInputStream fis;
        String svgContent = null;
        fis = new FileInputStream(path);
        svgContent = IOUtils.toString(fis, StandardCharsets.UTF_8);
        fis.close();
        return svgContent;
    }

    private void download(String fileName) throws IOException {
        LocalStore localStore = new LocalStore("");
        localStore.get(svgTemplate, fileName, directory);
    }

    private String getSvgFileName() {
        String fileName = null;
        try {
            URI uri = new URI(svgTemplate);
            String path = uri.getPath();
            fileName = path.substring(path.lastIndexOf('/') + 1);
            if (!fileName.endsWith(".svg"))
                return fileName.concat(".svg");
        } catch (URISyntaxException e) {
            logger.debug("Exception while getting file name from template url : {}", e.getMessage());
        }
        return fileName;
    }


}

File: certificate-processor/src/main/scala/org/sunbird/incredible/processor/views/SvgGenerator.scala
package org.sunbird.incredible.processor.views

import java.io.IOException
import java.util.regex.Matcher

import com.twitter.storehaus.cache.{Cache, LRUCache}
import org.apache.commons.lang.StringUtils
import org.apache.commons.text.StringSubstitutor
import org.slf4j.{Logger, LoggerFactory}
import org.sunbird.incredible.pojos.ob.CertificateExtension

import scala.io.{BufferedSource, Source}
import scala.util.matching.Regex

object EncoderMap {
  val encoder: Map[String, String] = Map("<" -> "%3C", ">" -> "%3E", "#" -> "%23", "%" -> "%25", "\"" -> "\'")
}


object SvgGenerator {
  private val logger: Logger = LoggerFactory.getLogger(getClass.getName)
  private var svgTemplatesCache: Cache[String, String] = LRUCache[String, String](15)

  @throws[IOException]
  def generate(certificateExtension: CertificateExtension, encodedQrCode: String, svgTemplateUrl: String): String = {
    var cachedTemplate = svgTemplatesCache.get(svgTemplateUrl).getOrElse("")
    if (StringUtils.isEmpty(cachedTemplate)) {
      logger.info("{} svg not cached , downloading", svgTemplateUrl)
      cachedTemplate = download(svgTemplateUrl)
      cachedTemplate = "data:image/svg+xml," + encodeData(cachedTemplate)
      cachedTemplate = cachedTemplate.replaceAll("\n", "").replaceAll("\t", "")
      svgTemplatesCache = svgTemplatesCache.put(svgTemplateUrl, cachedTemplate)._2
    } else {
      svgTemplatesCache = svgTemplatesCache.hit(svgTemplateUrl)
    }
    val svgData = replaceTemplateVars(cachedTemplate, certificateExtension, encodedQrCode)
    logger.info("svg template string creation completed")
    svgData
  }


  private def replaceTemplateVars(svgContent: String, certificateExtension: CertificateExtension, encodeQrCode: String): String = {
    val varResolver = new VarResolver(certificateExtension)
    val certData: java.util.Map[String, String] = varResolver.getCertMetaData
    certData.put("qrCodeImage", "data:image/png;base64," + encodeQrCode)
    println(certData)
    val sub = new StringSubstitutor(certData)
    val resolvedString = sub.replace(svgContent)
    logger.info("replacing temp vars completed")
    resolvedString
  }

  private def encodeData(data: String): String = {
    val stringBuffer: StringBuffer = new StringBuffer
    val regex: Regex = "[<>#%\"]".r
    val pattern: java.util.regex.Pattern = regex.pattern
    val matcher: Matcher = pattern.matcher(data)
    while (matcher.find)
      matcher.appendReplacement(stringBuffer, EncoderMap.encoder(matcher.group))
    matcher.appendTail(stringBuffer)
    stringBuffer.toString
  }

  @throws[IOException]
  private def download(svgTemplate: String): String = {
    var svgData: BufferedSource = null
    try {
      svgData = Source.fromURL(svgTemplate)
    } finally {
      svgData.close
    }
    svgData.mkString
  }

}
